P5-217: Added more about RTS (#139)
aau-hcl-p5/main: report/4.Design/RTS.tex

\section{Real-time system}\label{Design:RTS}
In order for the project to qualify as a real-time system it must be able to respond correctly in a timely manner.
In the following section, the design decisions made for the project in regards to ensuring this will be elaborated upon.

\subsection{Scheduling}\label{Design:Scheduling}
When choosing a scheduling method for the system, multiple methods were considered.

> Both fixed-priority scheduling (FPS) and cyclic executive (CE) were considered for the system as both of them would be good at handling the tasks of the system with their deadlines.

Reviewer: Abbreviation

\section{Real-time system}\label{Design:RTS}
In order for the project to qualify as a real-time system it must be able to respond correctly in a timely manner.
In the following section, the design decisions made for the project in regards to ensuring this will be elaborated upon.

\subsection{Scheduling}\label{Design:Scheduling}
When choosing a scheduling method for the system, multiple methods were considered.
Both fixed-priority scheduling (FPS) and cyclic executive (CE) were considered for the system as both of them would be good at handling the tasks of the system with their deadlines.

At first it seemed obvious to use the CE model, as all tasks should run regularly, and they were executed fast enough to run within the major cycle
In this case, the CE method is the perfect solution.
However, should the execution time of a single task exceed one millisecond, the usage of CE is no longer viable, as the task to keep the USB connection alive needs to be executed once every millisecond, as required by nxtOSEK.
In that case, FPS is a better solution, as it allows the scheduler to pre-empt the task in order to run the USB related task.

> Both the fixed priority and CE methods are a common ways to implement hard real-time systems due to their simplicity, deterministic behavior and relative ease of implementation\cite{CyclicExecutionKimLarsen}.

Reviewer 2: common way**~s~**

> In order to ensure that the system is schedulable, an UPPAAL model of the scheduler and the tasks was created.

Reviewer 2: tasks **~was~** were

Author: nej, objekt er an uppaal model

\subsubsection{UPPAAL}
UPPAAL is an integrated tool environment for modeling, validation and verification of real-time systems modeled as networks of timed automata\cite{UPPAALWebsite}.
Using the UPPAAL, a model of the tasks of the system was constructed.

\paragraph{Templates}

> When creating an UPPAAL project, a templates must be made.

Reviewer 2: template**~s~**

On edges different methods can be used to enforce a wanted behavior and transfer information between templates.

\paragraph{Declarations}
Once the templates has been created, the declaration is used to verify that the entire system is functional.
The declaration can contain variables, channels and methods used within templates and between templates.
Declarations can be specified per template or globally.

\paragraph{System declarations}
The system model in the templates is initialized from the system declarations, where instances of the templates are initialized and added to the system as can be seen on \autoref{snip:UPPAALSystemDeclaration}.

\begin{lstlisting}[label={snip:UPPAALSystemDeclaration},caption={System declaration from UPPAAL},frame=tlrb,numbers=none]
BackgroundTask    = Task( 0 );
UpdateDisplayTask = Task( 1 );
MoveMotorsTask    = Task( 2 );
ReciveDataTask    = Task( 3 );
HandleLaserTask   = Task( 4 );
KeepUSBAliveTask  = Task( 5 );

system BackgroundTask, UpdateDisplayTask, MoveMotorsTask, ReciveDataTask, HandleLaserTask, KeepUSBAliveTask, Scheduler;
\end{lstlisting}

> Each task in initialized and assigned to a new task.

Reviewer 2: something is missing, no?

The task is added to the system, along with the scheduler.

The behavior of a task is shown on \autoref{fig:UppaalTask}.

> The model is a timed automaton which is used to ensure that the task will come true to its deadline and periods.

Reviewer 2: Is come true the valid way to say this? Comply maybe?

The model consists of all the states a task can be in.
Each state has a number of transitions that can change the state of the task.

\figur{0.5}{images/Task}{Example of a task template from UPPAAAL}{fig:UppaalTask}

On \autoref{fig:UppaalScheduler} the model of how the scheduler works can be seen.

> Since the scheduler is responsible for running all tasks on the system it simply needs an queue of tasks ready for running.

Reviewer 2: a**~n~** queue

The tasks in the queue are sorted, based on their priority upon entering the queue.

> It when simply needs to start the tasks in the queue.

Reviewer 2: t**~w~**hen

\figur{0.5}{images/Scheduler}{The figure shows the scheduler template from UPPAAL}{fig:UppaalScheduler}

> Based on the model of both the task and the scheduler, it is verifiable that the system is schedulable using the verifier tool in UPPAAL.

Reviewer 2: verifier -> verification maybe?

When choosing a scheduling method for the system, multiple methods were considered. 
Both fixed priority scheduling and cyclic executive were considered for the system as both of them would be good at handling the tasks of the system with their deadlines.

> Due to the fast execution time of the tasks, and that preemption and context switching would exceed the time for simply completing the previous task, it was at first deemed obvious to use the cyclic executive model as all tasks should run regularly, and that they execute fast enough to run within the major cycle the cyclic execution method should be perfect. 

Reviewer 3: Lang sætning.. Consider rephrasing.

Reviewer 3: Især sidste den bliver meget svær at forstå  fra "executive model as..." og til slut. Der mangler som minimum komma "major cycle, the cyclic execution"

> When choosing a scheduling method for the system, multiple methods were considered. 

Reviewer: Generelt, hvis vi brugte cyclic executive og  FPS forkortelserne tidligere, skal den forkortelse bruges konsekvent

Both fixed priority scheduling and cyclic executive were considered for the system as both of them would be good at handling the tasks of the system with their deadlines.

Due to the fast execution time of the tasks, and that preemption and context switching would exceed the time for simply completing the previous task.
At first it was at deemed obvious to use the cyclic executive model as all tasks should run regularly, and that they execute fast enough to run within the major cycle the cyclic execution method should be perfect. 
However, should the execution time of a single task exceed one millisecond, the usage of cyclic executive is no longer viable, as the task to keep the USB connection alive is required to be executed once every millisecond.

> In this case, fixed-priority scheduling is a better alternative, as it allows the scheduler to pre-empt the task in order to run the USB related task.

Reviewer: ```suggestion
In that case, fixed-priority scheduling is a better solution, as it allows the scheduler to pre-empt the task in order to run the USB related task.
```

> Both the fixed priority and cyclic executive methods are a common ways to implement hard real-time systems due to their simplicity, deterministic behavior and relative ease of implementation\cite{CyclicExecutionKimLarsen}.

Reviewer: ```suggestion
Both the fixed priority and cyclic executive methods are common ways to implement hard real-time systems due to their simplicity, deterministic behavior and relative ease of implementation\cite{CyclicExecutionKimLarsen}.
```

Reviewer: Den her sætning generelt hvis også?

In order to ensure that the system is schedulable, an UPPAAL model of the scheduler and the tasks was created. 

\subsubsection{UPPAAL}

> UPPAAL is an integrated tool environment for modeling, validation and verification of real-time systems modeled as networks of timed automata\cite{UPPAALWebsite}. 

Reviewer: Er det ikke nok bare at sige: is a tool.

Reviewer: ```suggestion
UPPAAL is a tool for modeling, validating and verifying real-time systems modeled as networks of timed automata\cite{UPPAALWebsite}. 
```

> Using the UPPAAL tool, a model of the tasks of the system was constructed. 

Reviewer: Såfremt overstående var forkert.
```suggestion
Using the UPPAAL environment a model of the tasks of the system was constructed. 
```

Author: vil faktisk mene at der bare skal stå using UPPAAL @thomaslhansen ?

Reviewer: Constructed er også mærkeligt

\paragraph{Templates}

> When creating an UPPAAL project, a templates must be made, which is a model that describes a system or a part of a system.  

Reviewer: ```suggestion
When creating an UPPAAL project, a templates must be made.
Templates are a model that describes a system or a part of a system.  
```

Reviewer: Burde det ikke være part of a system, siden der nedenunder skrives at de er brugt til at lave et helt system.

> The templates are later used to make up the entire system, and they can communicate using channels specified in the declarations of the project.

Reviewer: Føler der er redundancy.
```suggestion
The templates can communicate using channels specified in the declarations of the project.
```

> 

Reviewer: Tror godt jeg kunne tænke mig lidt mere consistency mellem hvad der er en template og hvad der er en model. føler lidt det er lige meget at nævne at noget er en template når det er en model og omvendt.

\paragraph{Locations}
An UPPAAL model consists of at least one \textit{location}, which is a state that the system can be in. 

> A location can optionally be specified as \textit{initial}, \textit{urgent}, \textit{committed}. 

Reviewer: Måske, det står under?
```suggestion
A location can optionally be marked as \textit{initial}, \textit{urgent}, \textit{committed}. 
```

> If a location is marked as initial, it is defined as the starting state for the modelled system. 

Reviewer: Vi skriver amerikansk
```suggestion
If a location is marked as initial, it is defined as the starting state for the modeled system. 
```

A location marked with urgent will be the preferred destination, if there are multiple possible transitions. 

> A location marked as committed is an location that is preferred over an location without any type and that is overruled by a location marked as urgent. 

Reviewer: ```suggestion
A location marked as committed is a location that is preferred over locations without any type and is overruled by a location marked as urgent. 
```

> If a location is not marked with an specific type, the location might be taken but has the lowest priority if any other transitions are possible. 

Reviewer: ```suggestion
If a location is not marked with a specific type, the location has the lowest priority if any other transitions are possible. 
```

> A location can be given an \textit{invariant} that can be used to specify a condition that must be fulfilled before a location can be left. 

Reviewer: ```suggestion
A location can be given an \textit{invariant}, which specifies a condition that must be fulfilled before the location can be transitioned from.
```

\paragraph{Edges}

> Between locations an edge can be created, they are used to model a shift in model state and is used to control the flow through the template. 

Reviewer: ```suggestion
Between locations are edges.
Edges are used to model a shift in model state and is used to control the flow through the template. 
```

Author: måske slå dem sammen

between locations are edges, which are used to mode....

> On edges different methods can be used to enforce a wanted behavior and transfer information between templates. 

Reviewer: ```suggestion
Edges can have methods, which enforce a wanted behavior and transfer information between templates. 
```

Author: er ikke sikker på at det er samme mening - der kan bruges forskellige metoder til at håndtere opførsel af kanter

\paragraph{Declarations}

> Once the templates has been created, the declaration is used to make the entire system function. 

Reviewer: Also, er det ikke mere noget i stil med: ....to verify the modeled system.
```suggestion
Once the templates has been created, the declaration is used to make the entire system function. 
```

> The declaration is can contain variables, channels and methods used in the templates and between templates. 

Reviewer: ```suggestion
The declaration can contain variables, channels and methods used within templates and between templates. 
```

> Declarations can be specified on a template level and on an global level. 

Reviewer: ```suggestion
Declarations can be specified per template and globally.
```

Author: or globally?

\paragraph{System declarations}

> The system model in the templates is initialized from the system declarations, where instances of the templates are initialized and added to the system as can be seen on \autoref{snip:UPPAALSystemDeclaration}. 

Reviewer: ```suggestion
The system model in the templates, are initialized from the system declarations.
Instances of the templates are initialized and added to the system as shown in \autoref{snip:UPPAALSystemDeclaration}. 
```
fatter stadig ikke in vs on, tror nok det er in, når det er en snippet, on når det er et billed.
For det er "på billedet", men "i snippeten"

\begin{lstlisting}[label={snip:UPPAALSystemDeclaration},caption={System declaration from UPPAAL},frame=tlrb,numbers=none]
BackgroundTask    = Task( 0 );
UpdateDisplayTask = Task( 1 );
MoveMotorsTask    = Task( 2 );
ReciveDataTask    = Task( 3 );
HandleLaserTask   = Task( 4 );
KeepUSBAliveTask  = Task( 5 );

system BackgroundTask, UpdateDisplayTask, MoveMotorsTask, ReciveDataTask, HandleLaserTask, KeepUSBAliveTask, Scheduler;
\end{lstlisting}

> Each task in initialized and assigned to an new task which is then added to the system along with the scheduler. 

Reviewer: Also hvor fuck kom scheduler fra?
```suggestion
Each task in initialized and assigned to a new task.
The task is added to the system, along with the scheduler. 
```

> On \autoref{fig:UppaalTask} the behavior of the a task in can be seen. 

Reviewer: ```suggestion
The behavior of a task is shown on \autoref{fig:UppaalTask}.
```

> The model is a timed automaton which is used to ensure that the task will come true to its deadline and periods. 

Reviewer: ```suggestion
The model is a timed automaton, which is used to ensure that the task will fulfil its deadline and periods. 
```

Author: tror faktisk det skal være be true to its deadline

> The model is a timed automaton which is used to ensure that the task will come true to its deadline and periods. 

Reviewer: Faktisk føler jeg at de næste 3 linier (61, 62, 63 her) er redundant????

The model consists of all the states a task can be in.
Each state has a number of transitions that can change the state of the task. 

> \figur{0.5}{images/Task}{The figure shows the the task template from UPPAAAL}{fig:UppaalTask}

Reviewer: ```suggestion
\figur{0.5}{images/Task}{Example of a task template from UPPAAAL}{fig:UppaalTask}
```

> On \autoref{fig:UppaalScheduler} the model of how the scheduler works can be seen. 

Reviewer: Er lidt i tvivl om den her, har virkeligt stadig ikke forstået Upaal?

> Since the scheduler is responsible for running all tasks on the system it simply needs an queue of tasks ready for running. 

Reviewer: ```suggestion
As the scheduler is responsible for running all tasks on the system, it needs a queue of tasks that are ready to run. 
```

> The tasks in the queue are sorted then entering the queue based on their priority. 

Reviewer: ```suggestion
The tasks in the queue are sorted, based on their priority. 
```

> The tasks in the queue are sorted then entering the queue based on their priority. 

Reviewer: The tasks in the queue are sorted based on their priority.

> It when simply needs to start the tasks as they arrive at a ready state and the scheduler ready to run them.

Reviewer: hjælp

\figur{0.5}{images/Scheduler}{The figure shows the scheduler template from UPPAAL}{fig:UppaalScheduler}

Based on the model of both the task and the scheduler, it is verifiable that the system is schedulable using the verifier tool in UPPAAL. 

> The verifier will run a query, as can be seen on \autoref{snip:UPPAALQuery} asking that for any given state, none of the tasks are going into the error state. 

Reviewer: ```suggestion
The verifier will run a query, asking that for any given state, none of the tasks are going into the error state. 
This is shown on \autoref{snip:UPPAALQuery}.
```
